server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,650 @@
+                )
+            elif all(
+                [
+                    os.path.isfile(item),
+                    not base_item.endswith(".tar.xz"),
+                    not base_item.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_controllers(
+                    "unexpected_objects", controller, base_item
+                )
+            elif os.path.isfile(item) and (
+                base_item.endswith(".tar.xz") or base_item.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller)
+            else:
+                print(
+                    f"{base_item} item should have been handled by the above mentioned conditions. "
+                    f"It is an unexpected item which should not have occured, "
+                    f"leading to an inappropriate condition"
+                )
+        verify_subdirs(hier, controller, controller_subdir)
+        verify_prefixes(hier, controller)
+
+    return
+
+
+# Incoming Hierarchy
+def verify_tar_dirs(ihier, tarball_dirs, tblist, controller):
+    for tb in tarball_dirs:
+        if tb.endswith("unpack"):
+            tar = tb[:-7]
+            tar = f"{tar}.tar.xz"
+            val = "invalid_unpacking_dirs"
+        else:
+            tar = f"{tb}.tar.xz"
+            val = "invalid_tb_dirs"
+        tarfile = os.path.join(ihier.config.ARCHIVE, controller, tar)
+        if os.path.exists(tarfile):
+            with open(tarfile, "r") as f:
+                try:
+                    file_content = f.read(1)
+                    if file_content:
+                        continue
+                except Exception:
+                    pass
+        else:
+            tblist(val, controller, os.path.basename(tb))
+    return
+
+
+# 'ihier' is IncomingHierarchy class object
+def verify_incoming(ihier, verifylist, hierarchy=False):
+
+    for controller in verifylist:
+        ihier.add_controller(controller)
+        direct_entries = glob.iglob(
+            os.path.join(ihier.config.INCOMING, controller, "*")
+        )
+        tarball_dirs, unpacking_tarball_dirs = (list() for i in range(2))
+
+        if not os.path.isdir(os.path.join(ihier.config.ARCHIVE, controller)):
+            # Skip incoming controller directories that don't have an $ARCHIVE
+            # directory, handled in another part of the audit.
+            continue
+
+        for dirent in direct_entries:
+            dirt = os.path.basename(dirent)
+            if (
+                os.path.isdir(dirent)
+                and not dirt.endswith(".unpack")
+                and not len(os.listdir(dirent)) == 0
+            ):
+                tarball_dirs.append(dirt)
+            elif (
+                os.path.isdir(dirent)
+                and not dirt.endswith(".unpack")
+                and len(os.listdir(dirent)) == 0
+            ):
+                ihier.add_unexpected_controllers("empty_tarball_dirs", controller, dirt)
```

Comment:
Again, the propagation of these literal strings "all over the place" seems potentially risky; typos are too easy, and may be hard to catch. Even with thorough unit testing, having centralized constants like Hierarchy.EMPTY_TARBALL_DIRS would make errors much easier to catch during editing. (As well as allowing autocomplete to help.)

Reply:
It seems to be a good call. I wanted to clear things on my end as I got a little confused.

Current implementation:
```
class ArchiveHierarchy(Hierarchy):
    def __init__(self, name, path, config):
        super().__init__(name, path, config)

        self.validation_list = {
            "unexpected_dirs": DictOfList(
                "Unexpected state directories found in this controller directory:"
            ),
            "subdir_status_indicators": DictOfList(
                "No state directories found in this controller directory."
            ),
            "unexpected_symlinks": DictOfList(
                "Unexpected symlinks in controller directory:"
            ),
```

What I understood:
```
class Hierarchy:
    def __init__(self, name, path, config):

        self.UNEXPECTED_DIRS = DictOfList(
                "Unexpected state directories found in this controller directory:"
         ),
         self.SUBDIR_STATUS_INDICATORS = DictOfList(
                "No state directories found in this controller directory."
         ),
         self.UNEXPECTED_SYMLINKS = DictOfList(
                "Unexpected symlinks in controller directory:"
         )
```
And we should specify each validation_list key like this?

Reply:
That's fine, and even has some advantages; however what I'd had in mind would retain the dict implementation you've got now but formalize the keys with something like

```
class Hierarchy:
  UNEXPECTED_DIRS = "unexpected_dirs"
  UNEXPECTED_SYMLINKS = "unexpected_symlinks"

  self.validation_list = {}
}

Class ArchiveHierarchy(Hierarchy):
  super().__init__(name, path, config)

  self.validation_list = {
    UNEXPECTED_DIRS: DictOfList(...),
    UNEXPECTED_SYMLINKS: DictOfList(...),
    ...
  }
}
```

(Or even just `self.validation_list[UNEXPECTED_DIRS] = DictOfList()`, since the superclass is ensuring the dict already exists; we don't need to replace it.)

But I think the main practical difference is that if your subclasses don't all use the same keys, their keysets won't need to include DictOfList objects they're not using. Separate object properties as you proposed formalizes the documentation of the individual lists, which is a good thing, at the expense of some potentially wasted space.

Reply:
ok